File: dart/1298.dart
class Solution {
    int traverse(List<int> s, List<int> c, List<List<int>> k, List<List<int>> b, int j, List<bool> p, List<bool> v) {
        if (s[j] == 0) {
            p[j] = true;
            return 0;
        }
        if (v[j])
            return 0;
        v[j] = true;
        int r = c[j];
        for (int m in k[j]) {
            s[m] = 1;
            if (p[m])
                r += traverse(s, c, k, b, m, p, v);
        }
        for (int m in b[j])
            r += traverse(s, c, k, b, m, p, v);
        return r;
    }

    int maxCandies(List<int> s, List<int> c, List<List<int>> k, List<List<int>> b, List<int> i) {
        int n = s.length;
        List<bool> p = [];
        List<bool> v = [];
        for (int j = 0; j < n; j++) {
            p.add(false);
            v.add(false);
        }
        int r = 0;
        for (int j in i)
            r += traverse(s, c, k, b, j, p, v);
        return r;
    }
}

File: python3/1298.py
class Solution:
    def maxCandies(
        self,
        s: List[int],
        c: List[int],
        k: List[List[int]],
        b: List[List[int]],
        i: List[int],
    ) -> int:
        n = len(s)
        p = [False] * n
        v = [False] * n

        def traverse(j: int) -> int:
            if s[j] == 0:
                p[j] = True
                return 0
            if v[j]:
                return 0
            v[j] = True
            r = c[j]
            for m in k[j]:
                s[m] = 1
                if p[m]:
                    r += traverse(m)
            for m in b[j]:
                r += traverse(m)
            return r

        r = 0
        for j in i:
            r += traverse(j)
        return r
